File: scala/revenj-core/src/main/scala/net/revenj/database/postgres/converters/ShortConverter.scala
package net.revenj.database.postgres.converters

import net.revenj.database.postgres.{PostgresBuffer, PostgresReader, PostgresWriter}

object ShortConverter extends Converter[Short] {
  override def serializeURI(sw: PostgresBuffer, value: Short): Unit = {
    val offset = NumberConverter.serialize(value, sw.tempBuffer)
    sw.addToBuffer(sw.tempBuffer, offset, 6)
  }

  override val dbName = "int2"

  override def default(): Short = 0

  override def parseRaw(reader: PostgresReader, start: Int, context: Int): Short = parseShort(reader, start, ')')

  private def parseShort(reader: PostgresReader, start: Int, matchEnd: Char): Short = {
    var res = 0
    var cur = start
    if (cur == '-') {
      cur = reader.read()
      do {
        res = (res << 3) + (res << 1) - (cur - '0')
        cur = reader.read()
      } while (cur != -1 && cur != ',' && cur != matchEnd)
    } else {
      do {
        res = (res << 3) + (res << 1) + (cur - '0')
        cur = reader.read()
      } while (cur != -1 && cur != ',' && cur != matchEnd)
    }
    res.toShort
  }

  override def parseCollectionItem(reader: PostgresReader, context: Int): Short = {
    val cur = reader.read()
    if (cur == 'N') {
      reader.read(4)
      0
    } else {
      parseShort(reader, cur, '}')
    }
  }

  override def parseNullableCollectionItem(reader: PostgresReader, context: Int): Option[Short] = {
    val cur = reader.read()
    if (cur == 'N') {
      reader.read(4)
      None
    } else {
      Some(parseShort(reader, cur, '}'))
    }
  }

  override def toTuple(value: Short): PostgresTuple = IntConverter.toTuple(value)
}

File: java/revenj-core/src/main/java/org/revenj/database/postgres/converters/ShortConverter.java
package org.revenj.database.postgres.converters;

import org.revenj.database.postgres.PostgresBuffer;
import org.revenj.database.postgres.PostgresReader;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public abstract class ShortConverter {

	public static void serializeURI(PostgresBuffer sw, short value) throws IOException {
		int offset = NumberConverter.serialize(value, sw.getTempBuffer());
		sw.addToBuffer(sw.getTempBuffer(), offset, 6);
	}

	public static void serializeURI(PostgresBuffer sw, Short value) throws IOException {
		if (value == null) return;
		serializeURI(sw, value.shortValue());
	}

	public static Short parseNullable(PostgresReader reader) {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return null;
		}
		return parseShort(reader, cur, ')');
	}

	public static short parse(PostgresReader reader) {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return 0;
		}
		return parseShort(reader, cur, ')');
	}

	private static short parseShort(PostgresReader reader, int cur, char matchEnd) {
		int res = 0;
		if (cur == '-') {
			cur = reader.read();
			do {
				res = (res << 3) + (res << 1) - (cur - 48);
				cur = reader.read();
			} while (cur != -1 && cur != ',' && cur != matchEnd);
		} else {
			do {
				res = (res << 3) + (res << 1) + (cur - 48);
				cur = reader.read();
			} while (cur != -1 && cur != ',' && cur != matchEnd);
		}
		return (short) res;
	}

	public static List<Short> parseCollection(PostgresReader reader, int context, boolean allowNulls) {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return null;
		}
		boolean escaped = cur != '{';
		if (escaped) {
			reader.read(context);
		}
		cur = reader.peek();
		if (cur == '}') {
			if (escaped) {
				reader.read(context + 2);
			} else {
				reader.read(2);
			}
			return new ArrayList<>(0);
		}
		Short defaultValue = allowNulls ? null : (short)0;
		List<Short> list = new ArrayList<>();
		do {
			cur = reader.read();
			if (cur == 'N') {
				list.add(defaultValue);
				cur = reader.read(4);
			} else {
				list.add(parseShort(reader, cur, '}'));
				cur = reader.last();
			}
		} while (cur == ',');
		if (escaped) {
			reader.read(context + 1);
		} else {
			reader.read();
		}
		return list;
	}

	public static PostgresTuple toTuple(Short value) {
		return value == null ? null : IntConverter.toTuple(value);
	}

	public static PostgresTuple toTuple(short value) {
		return IntConverter.toTuple(value);
	}
}
